cmd/ingest/main.go

```diff
@@ -90,117 +105,158 @@ func main() {
 		}).Fatal("couldn't create SQL tables")
 	}
 
-	// Initialize ZMQ
-	ctx, err := zmq.NewContext()
+	//Initialize RPC connection with full node zcashd
+	rpcClient, err := frontend.NewZRPCFromConf(opts.zcashConfPath)
 	if err != nil {
 		log.WithFields(logrus.Fields{
 			"error": err,
-		}).Fatal("couldn't create zmq context")
+		}).Warn("zcash.conf failed, will try empty credentials for rpc")
+
+		//Default to testnet, but user MUST specify rpcuser and rpcpassword in zcash.conf; no default
+		rpcClient, err = frontend.NewZRPCFromCreds(opts.rpcHost+":"+opts.rpcPort, opts.rpcLogin, opts.rpcPass)
+
+		if err != nil {
+			log.WithFields(logrus.Fields{
+				"error": err,
+			}).Fatal("couldn't start rpc connection")
+		}
 	}
-	defer ctx.Term()
 
-	// WARNING: The Socket is not thread safe. This means that you cannot
-	// access the same Socket from different goroutines without using something
-	// like a mutex.
-	sock, err := ctx.NewSocket(zmq.SUB)
+	ctx := context.Background()
+	height, err := storage.GetCurrentHeight(ctx, db)
 	if err != nil {
 		log.WithFields(logrus.Fields{
 			"error": err,
-		}).Fatal("couldn't create zmq context socket")
-	}
+  	}).Warn("unable to get current height from local db storage")
 
-	err = sock.SetSubscribe(opts.zmqTopic)
-	if err != nil {
-		log.WithFields(logrus.Fields{
-			"error":  err,
-			"stream": opts.zmqTopic,
-		}).Fatal("couldn't subscribe to stream")
 	}
 
-	connString := fmt.Sprintf("tcp://%s", opts.zmqAddr)
-
-	err = sock.Connect(connString)
-	if err != nil {
+	//ingest from Sapling testnet height
+	if height < 419200 {
+		height = 419200
 		log.WithFields(logrus.Fields{
-			"error":      err,
-			"connection": connString,
-		}).Fatal("couldn't connect to socket")
+			"error": err,
+  	}).Warn("invalid current height read from local db storage")
 	}
-	defer sock.Close()
 
-	log.Printf("Listening to 0mq on %s", opts.zmqAddr)
+	if opts.startHeight != 0 {
+		height = int(opts.startHeight)
+	}
 
+	timeout_count := 0
 	// Start listening for new blocks
 	for {
-		msg, err := sock.RecvMessageBytes(0)
+		block, err := getBlock(rpcClient, height)
 		if err != nil {
 			log.WithFields(logrus.Fields{
-				"error": err,
-			}).Error("error on msg recv")
-			continue
+				"height": height,
+				"error":  err,
+			}).Warn("error with getblock")
+			timeout_count++
+			if timeout_count == 3 {
+				log.WithFields(logrus.Fields{
+					"timeouts":  timeout_count,
+				}).Warn("unable to issue RPC call to zcashd node 3 times")
+				timeout_count = 0
+				time.Sleep(3 * 60 * time.Second)
+			}
 		}
-
-		if len(msg) < 3 {
-			log.WithFields(logrus.Fields{
-				"msg": msg,
-			}).Warn("got unknown message type")
-			continue
+		if block != nil {
+			handleBlock(db, block)
+			height++
+			if timeout_count > 0 {
+				timeout_count--
+			}
+			//TODO store block current/prev hash for formal reorg
+		} else {
+			height -= 4
```

Comment:
Rewinding by 4 blocks every time we reach this point seems like it would cause every server to ingest over 4 times more data from zcashd because depending on the block time, this could rewind multiple times for each block. We may want to find a more efficient way to address the problem that this is trying to solve, instead.